### DISCUSSION TITLE
Complex component needs to receive and update multiple fields

### DISCUSSION THREAD

**Discussion by DirkWolthuis** [react-hook-form/react-hook-form#3518] [Q&A] [answered]
Hi all, I'm currently refactoring an app with react-hook-form (👋 Formik) and I have a problem. A complex component needs to receive multiple form values and updates multiple values. I have made a [codesandbox ](https://codesandbox.io/s/react-hook-form-complex-component-update-more-than-1-field-rrxs1?file=/src/index.js) that demonstrates the problem. If I set defaultValues that I later need (a couple components down in a tree) than those values are always undefined with the `getValue` function (because they aren't registered I guess). Is there a way to make this work for my complex component? So multiple values in and multiple values out (with the `setValue` function I guess). 

The code (in simple form):
```javascript
import React from "react";
import ReactDOM from "react-dom";
import { useForm, Controller } from "react-hook-form";
import "./styles.css";

const defaultValues = {
  name: "Dirk",
  ids: {
    id1: 123,
    id2: 8322,
    id3: 9819203
  }
};

function App() {
  const { handleSubmit, errors, control, setValue, getValues } = useForm({
    defaultValues: defaultValues
  });
  const onSubmit = (data) => {
    //console.log(data);
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <label>FirstName</label>

      <Controller as={<input />} name="name" control={control} />

      <label>IDS</label>

      {errors.LastName && <p>This field is required</p>}
      <ComplexComponent
        ids={getValues("ids")}
        onChange={(checkObjectIds) => {
          Object.keys(checkObjectIds).forEach((property) => {
            setValue(property, checkObjectIds[property]);
          });
        }}
      />
      <input type="submit" />
    </form>
  );
}

const ComplexComponent = ({ ids, onChange }) => {
  //complex stuf
  console.log(ids);
  return (
    <button
      onClick={() => {
        onChange({ id1: ids.id1 + 1, id2: ids.id2 + 1, id3: ids.id3 + 1 });
      }}
    >
      Do something complex
    </button>
  );
};

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);

```

Thanks!

**Comment by bluebill1049** [comment, marked as answer]
you can use `useWatch` to watch other fields.

https://react-hook-form.com/api#useWatch

To update other inputs, yes you are right with `setValue`

**Reply by DirkWolthuis** [reply]
> you can use `useWatch` to watch other fields.
> 
> https://react-hook-form.com/api#useWatch
> 
> To update other inputs, yes you are right with `setValue`

Hi Bill, 

I've updated the [codesandbox](https://codesandbox.io/s/react-hook-form-complex-component-update-more-than-1-field-rrxs1?file=/src/index.js) but the form values on submit don't contain the id-values. Can you look and see what I'm doing wrong?

**Reply by bluebill1049** [reply]
you update those input without registering them, but you can work around this issue with 

`shouldUnregister: false,` at useForm.

**Reply by DirkWolthuis** [reply]
Thank you very much!
